bitcoin/src/bip32.rs

```diff
@@ -295,6 +294,162 @@ impl serde::Serialize for ChildNumber {
+    /// const NORMAL: u32 = 0;
+    /// assert_eq!(path.to_u32_vec(), vec![84 + NORMAL, NORMAL, NORMAL, 0, 1]);
+    /// ```
+    pub fn to_u32_vec(self) -> Vec<u32> { self.into_iter().map(|num| num.0).collect() }
+
+    /// Creates a derivation path from a slice of u32s.
+    /// ```
+    /// use bitcoin::bip32::NormalDerivationPath;
+    ///
+    /// const NORMAL: u32 = 0;
+    /// let expected = vec![84 + NORMAL, NORMAL, NORMAL, 0, 1];
+    /// let path = NormalDerivationPath::from_u32_slice(expected.as_slice());
+    /// assert_eq!(path.to_u32_vec(), expected);
+    /// ```
+    pub fn from_u32_slice(numbers: &[u32]) -> Self {
+        numbers.iter().map(|&n| NormalChildNumber(n)).collect()
+    }
+}
+impl AsRef<[NormalChildNumber]> for NormalDerivationPath {
+    fn as_ref(&self) -> &[NormalChildNumber] { &self.0 }
+}
+
+impl From<NormalDerivationPath> for DerivationPath {
+    fn from(path: NormalDerivationPath) -> Self {
+        path.0.iter().map(|n| ChildNumber::from(*n)).collect()
+    }
+}
+
+impl FromIterator<NormalChildNumber> for NormalDerivationPath {
+    fn from_iter<T: IntoIterator<Item = NormalChildNumber>>(iter: T) -> Self {
+        NormalDerivationPath(Vec::from_iter(iter))
+    }
+}
+
+impl fmt::Display for NormalDerivationPath {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for (i, child) in self.0.iter().enumerate() {
+            if i != 0 {
+                write!(f, "/")?;
+            }
+            write!(f, "{}", child.0)?;
+        }
+        Ok(())
+    }
+}
+
+impl TryFrom<&[ChildNumber]> for NormalDerivationPath {
+    type Error = Error;
+    fn try_from(value: &[ChildNumber]) -> Result<Self, Self::Error> {
+        let mut base = Vec::with_capacity(value.len());
+        for child in value.iter() {
+            match child {
+                ChildNumber::Normal(number) => {
+                    base.push(*number);
+                }
+                ChildNumber::Hardened(_) => {
+                    return Err(Error::CannotDeriveFromHardenedKey);
+                }
+            }
+        }
+        Ok(NormalDerivationPath(base))
+    }
+}
+
+impl From<&[NormalChildNumber]> for NormalDerivationPath {
+    fn from(value: &[NormalChildNumber]) -> Self { NormalDerivationPath(value.to_vec()) }
+}
+
+impl From<Vec<NormalChildNumber>> for NormalDerivationPath {
+    fn from(value: Vec<NormalChildNumber>) -> Self { NormalDerivationPath(value) }
+}
+
+impl FromStr for NormalDerivationPath {
+    type Err = Error;
+
+    fn from_str(path: &str) -> Result<NormalDerivationPath, Error> {
+        if path.is_empty() || path == "m" || path == "m/" {
+            return Ok(vec![].into());
+        }
+
+        let path = path.strip_prefix("m/").unwrap_or(path);
+
+        let parts = path.split('/');
+        let ret: Result<Vec<ChildNumber>, Error> = parts.map(str::parse).collect();
+        Ok(ret?.as_slice().try_into()?)
+    }
+}
+
+impl IntoIterator for NormalDerivationPath {
```

Comment:
It'd be nice to also have an impl for references. (And also the `.iter()` method.)